File: mobile/flutter_app/lib/domain/entities/visits.dart
/// Farm Visits & Agri-Tourism module — local domain entities (tech spec
/// v0.6 §4). Field names mirror `backend/app/domain/visits_models.py` so
/// payloads round-trip cleanly once sync is wired to the real API.
///
/// The module license itself reuses [ModuleLicense] from
/// `domain/entities/mouneh.dart` — one generic license class backs every
/// licensed module on both the backend and here, exactly like
/// `mouneh_models.ModuleLicense` is reused for `module_code =
/// "visits_agritourism"` on the server.
library;

const String kVisitsModuleCode = 'visits_agritourism';

const List<String> kVisitSessionStatuses = ['open', 'full', 'closed', 'cancelled', 'completed'];
const List<String> kVisitBookingStatuses = ['draft', 'confirmed', 'checked_in', 'completed', 'cancelled', 'no_show', 'refunded'];
const List<String> kVisitBookingSources = ['manual', 'whatsapp', 'website', 'phone', 'walk_in'];
const List<String> kVisitActivityTypes = ['tour', 'ride', 'workshop', 'tasting', 'event', 'other'];
const List<String> kVisitCostCategories = ['staff', 'cleaning', 'utilities', 'tasting', 'marketing', 'safety', 'maintenance', 'other'];
const List<String> kVisitCostAllocationMethods = ['per_session', 'per_guest', 'per_package', 'per_activity'];
const List<String> kVisitRetailChannels = ['farm_shop', 'tasting_upgrade', 'delivery_after_visit'];
const List<String> kVisitIncidentTypes = ['safety', 'animal', 'weather', 'payment', 'complaint', 'other'];
const List<String> kVisitIncidentSeverities = ['low', 'medium', 'high'];

/// RULE-VIS-003: opening days are configurable per farm, never hard-coded.
class VisitOpeningCalendarDay {
  const VisitOpeningCalendarDay({
    required this.weekday, // 0=Monday .. 6=Sunday
    this.isOpen = false,
    this.openTime,
    this.closeTime,
    this.defaultCapacity = 0,
    this.notes,
  });

  final int weekday;
  final bool isOpen;
  final String? openTime; // "HH:MM"
  final String? closeTime;
  final int defaultCapacity;
  final String? notes;

  VisitOpeningCalendarDay copyWith({bool? isOpen, String? openTime, String? closeTime, int? defaultCapacity, String? notes}) => VisitOpeningCalendarDay(
        weekday: weekday,
        isOpen: isOpen ?? this.isOpen,
        openTime: openTime ?? this.openTime,
        closeTime: closeTime ?? this.closeTime,
        defaultCapacity: defaultCapacity ?? this.defaultCapacity,
        notes: notes ?? this.notes,
      );
}

class VisitSession {
  const VisitSession({
    required this.id,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.capacity,
    this.status = 'open',
    this.weatherNote,
    this.expectedStaffCost,
  });

  final String id;
  final DateTime date;
  final String startTime;
  final String endTime;
  final int capacity;
  final String status;
  final String? weatherNote;
  final double? expectedStaffCost;

  VisitSession copyWith({int? capacity, String? status, String? weatherNote, double? expectedStaffCost}) => VisitSession(
        id: id,
        date: date,
        startTime: startTime,
        endTime: endTime,
        capacity: capacity ?? this.capacity,
        status: status ?? this.status,
        weatherNote: weatherNote ?? this.weatherNote,
        expectedStaffCost: expectedStaffCost ?? this.expectedStaffCost,
      );
}

/// A sellable bundled experience — created dynamically through the
/// Package Builder; nothing here is a fixed catalog entry.
class VisitPackage {
  const VisitPackage({
    required this.id,
    required this.name,
    this.description,
    this.basePrice = 0,
    this.currency = 'USD',
    this.durationMinutes,
    this.active = true,
  });

  final String id;
  final String name;
  final String? description;
  final double basePrice;
  final String currency;
  final int? durationMinutes;
  final bool active;
}

/// An individual bookable activity — "Horse Ride" is only ever demo data
/// (see data/demo/visits_demo_data.dart); a manager can create any
/// activity through the Activity Manager screen.
class VisitActivity {
  const VisitActivity({
    required this.id,
    required this.name,
    this.activityType = 'other',
    this.price = 0,
    this.capacityPerSlot = 1,
    this.durationMinutes,
    this.requiresStaffRole,
    this.requiresAnimalId,
    this.maxUsesPerDay,
    this.active = true,
  });

  final String id;
  final String name;
  final String activityType;
  final double price;
  final int capacityPerSlot;
  final int? durationMinutes;
  final String? requiresStaffRole;
  final String? requiresAnimalId;
  final int? maxUsesPerDay; // welfare_limit_json.max_uses_per_day, flattened for the mobile model
  final bool active;
}

class VisitorProfile {
  const VisitorProfile({
    required this.id,
    required this.fullName,
    this.phone,
    this.email,
    this.preferredLanguage = 'en',
    this.notes,
    this.consentMarketing = false,
  });

  final String id;
  final String fullName;
  final String? phone;
  final String? email;
  final String preferredLanguage;
  final String? notes;
  final bool consentMarketing;
}

class VisitBookingActivity {
  const VisitBookingActivity({
    required this.activityId,
    required this.scheduledAt,
    this.quantity = 1,
    this.unitPrice = 0,
    this.status = 'scheduled', // scheduled | completed | cancelled | missed
  });

  final String activityId;
  final DateTime scheduledAt;
  final int quantity;
  final double unitPrice;
  final String status;

  VisitBookingActivity copyWith({String? status}) => VisitBookingActivity(
        activityId: activityId,
        scheduledAt: scheduledAt,
        quantity: quantity,
        unitPrice: unitPrice,
        status: status ?? this.status,
      );
}

/// A visitor reservation. Status transitions are validated by
/// `lib/visits/analytics.dart::validateStatusTransition` (RULE-VIS-008);
/// every transition is written as an event by `VisitsWriteService`, never
/// silently overwritten.
class VisitBooking {
  const VisitBooking({
    required this.id,
    required this.visitorId,
    required this.sessionId,
    required this.packageId,
    this.status = 'draft',
    this.adults = 1,
    this.children = 0,
    this.totalAmount = 0,
    this.depositAmount = 0,
    this.balanceDue = 0,
    this.source = 'manual',
    this.notes,
    this.activities = const [],
    required this.createdAt,
    this.confirmedAt,
    this.checkedInAt,
    this.completedAt,
    this.cancelledAt,
  });

  final String id;
  final String visitorId;
  final String sessionId;
  final String packageId;
  final String status;
  final int adults;
  final int children;
  final double totalAmount;
  final double depositAmount;
  final double balanceDue;
  final String source;
  final String? notes;
  final List<VisitBookingActivity> activities;
  final DateTime createdAt;
  final DateTime? confirmedAt;
  final DateTime? checkedInAt;
  final DateTime? completedAt;
  final DateTime? cancelledAt;

  int get guestCount => adults + children;
  bool get isDraft => status == 'draft';
  bool get isConfirmed => status == 'confirmed';
  bool get isCheckedIn => status == 'checked_in';

  VisitBooking copyWith({
    String? status,
    double? totalAmount,
    double? balanceDue,
    List<VisitBookingActivity>? activities,
    DateTime? confirmedAt,
    DateTime? checkedInAt,
    DateTime? completedAt,
    DateTime? cancelledAt,
  }) =>
      VisitBooking(
        id: id,
        visitorId: visitorId,
        sessionId: sessionId,
        packageId: packageId,
        status: status ?? this.status,
        adults: adults,
        children: children,
        totalAmount: totalAmount ?? this.totalAmount,
        depositAmount: depositAmount,
        balanceDue: balanceDue ?? this.balanceDue,
        source: source,
        notes: notes,
        activities: activities ?? this.activities,
        createdAt: createdAt,
        confirmedAt: confirmedAt ?? this.confirmedAt,
        checkedInAt: checkedInAt ?? this.checkedInAt,
        completedAt: completedAt ?? this.completedAt,
        cancelledAt: cancelledAt ?? this.cancelledAt,
      );
}

class VisitStaffRosterEntry {
  const VisitStaffRosterEntry({
    required this.id,
    required this.sessionId,
    required this.workerId,
    required this.workerName,
    required this.role,
    required this.startTime,
    required this.endTime,
    this.hourlyRate = 0,
    this.totalCost = 0,
  });

  final String id;
  final String sessionId;
  final String workerId;
  final String workerName;
  final String role;
  final String startTime;
  final String endTime;
  final double hourlyRate;
  final double totalCost;
}

class VisitCost {
  const VisitCost({
    required this.id,
    required this.sessionId,
    required this.category,
    this.description,
    this.amount = 0,
    this.allocationMethod = 'per_session',
  });

  final String id;
  final String sessionId;
  final String category;
  final String? description;
  final double amount;
  final String allocationMethod;
}

class VisitRetailSale {
  const VisitRetailSale({
    required this.id,
    this.bookingId,
    this.visitorId,
    required this.channel,
    required this.totalAmount,
    required this.soldAt,
  });

  final String id;
  final String? bookingId;
  final String? visitorId;
  final String channel;
  final double totalAmount;
  final DateTime soldAt;
}

class VisitorFeedbackEntry {
  const VisitorFeedbackEntry({
    required this.id,
    required this.bookingId,
    required this.rating,
    this.comments,
    this.wouldReturn,
    required this.submittedAt,
  });

  final String id;
  final String bookingId;
  final int rating;
  final String? comments;
  final bool? wouldReturn;
  final DateTime submittedAt;
}

class VisitIncident {
  const VisitIncident({
    required this.id,
    required this.sessionId,
    this.bookingId,
    required this.incidentType,
    this.severity = 'low',
    required this.description,
    this.actionTaken,
    required this.createdAt,
  });

  final String id;
  final String sessionId;
  final String? bookingId;
  final String incidentType;
  final String severity;
  final String description;
  final String? actionTaken;
  final DateTime createdAt;
}

/// Per-scope roll-up shown on the Profitability Report (tech spec v0.6 §9).
class VisitProfitability {
  const VisitProfitability({
    required this.packageRevenue,
    required this.activityRevenue,
    required this.retailRevenue,
    required this.visitorRevenue,
    required this.staffCost,
    required this.cleaningUtilitiesCost,
    required this.otherCost,
    required this.directVisitCost,
    required this.grossMargin,
    required this.checkedInVisitors,
    required this.revenuePerVisitor,
    required this.retailConversionPct,
    required this.averageBasketValue,
  });

  final double packageRevenue;
  final double activityRevenue;
  final double retailRevenue;
  final double visitorRevenue;
  final double staffCost;
  final double cleaningUtilitiesCost;
  final double otherCost;
  final double directVisitCost;
  final double grossMargin;
  final int checkedInVisitors;
  final double revenuePerVisitor;
  final double retailConversionPct;
  final double averageBasketValue;
}

File: backend/app/schemas/visits.py
"""Farm Visits & Agri-Tourism — request/response schemas (tech spec v0.6
§4 "Data Model" + §6 "API Endpoints")."""
from __future__ import annotations

from datetime import date, datetime, time

from pydantic import BaseModel, field_validator

from app.schemas.common import ORMModel

SESSION_STATUSES = {"open", "full", "closed", "cancelled", "completed"}
BOOKING_STATUSES = {"draft", "confirmed", "checked_in", "completed", "cancelled", "no_show", "refunded"}
BOOKING_SOURCES = {"manual", "whatsapp", "website", "phone", "walk_in"}
ACTIVITY_TYPES = {"tour", "ride", "workshop", "tasting", "event", "other"}
BOOKING_ACTIVITY_STATUSES = {"scheduled", "completed", "cancelled", "missed"}
COST_CATEGORIES = {"staff", "cleaning", "utilities", "tasting", "marketing", "safety", "maintenance", "other"}
COST_ALLOCATION_METHODS = {"per_session", "per_guest", "per_package", "per_activity"}
RETAIL_CHANNELS = {"farm_shop", "tasting_upgrade", "delivery_after_visit"}
INCIDENT_TYPES = {"safety", "animal", "weather", "payment", "complaint", "other"}
INCIDENT_SEVERITIES = {"low", "medium", "high"}


# ---------------------------------------------------------------------------
# Opening calendar (REQ-VIS: configurable, never hard-coded weekends)
# ---------------------------------------------------------------------------
class OpeningCalendarDayUpsert(BaseModel):
    weekday: int
    is_open: bool = False
    open_time: time | None = None
    close_time: time | None = None
    default_capacity: int = 0
    notes: str | None = None

    @field_validator("weekday")
    @classmethod
    def weekday_range(cls, v: int) -> int:
        if not (0 <= v <= 6):
            raise ValueError("weekday must be between 0 (Monday) and 6 (Sunday)")
        return v

    @field_validator("default_capacity")
    @classmethod
    def capacity_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("default_capacity cannot be negative")
        return v


class OpeningCalendarDayOut(ORMModel):
    id: str
    weekday: int
    is_open: bool
    open_time: time | None = None
    close_time: time | None = None
    default_capacity: int
    notes: str | None = None


# ---------------------------------------------------------------------------
# Sessions
# ---------------------------------------------------------------------------
class VisitSessionCreate(BaseModel):
    date: date
    start_time: time
    end_time: time
    capacity: int
    weather_note: str | None = None
    expected_staff_cost: float | None = None

    @field_validator("capacity")
    @classmethod
    def capacity_positive(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("capacity must be greater than zero")
        return v


class VisitSessionUpdate(BaseModel):
    capacity: int | None = None
    status: str | None = None
    weather_note: str | None = None
    expected_staff_cost: float | None = None

    @field_validator("status")
    @classmethod
    def status_valid(cls, v: str | None) -> str | None:
        if v is not None and v not in SESSION_STATUSES:
            raise ValueError(f"status must be one of {sorted(SESSION_STATUSES)}")
        return v


class VisitSessionOut(ORMModel):
    id: str
    farm_id: str
    date: date
    start_time: time
    end_time: time
    capacity: int
    status: str
    weather_note: str | None = None
    expected_staff_cost: float | None = None


# ---------------------------------------------------------------------------
# Packages
# ---------------------------------------------------------------------------
class VisitPackageCreate(BaseModel):
    name: str
    description: str | None = None
    base_price: float = 0
    currency: str = "USD"
    duration_minutes: int | None = None
    included_items_json: dict = {}
    active: bool = True

    @field_validator("name")
    @classmethod
    def name_non_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("name cannot be empty")
        return v.strip()

    @field_validator("base_price")
    @classmethod
    def base_price_non_negative(cls, v: float) -> float:
        if v < 0:
            raise ValueError("base_price cannot be negative")
        return v


class VisitPackageOut(ORMModel):
    id: str
    farm_id: str
    name: str
    description: str | None = None
    base_price: float
    currency: str
    duration_minutes: int | None = None
    included_items_json: dict
    active: bool


# ---------------------------------------------------------------------------
# Activities
# ---------------------------------------------------------------------------
class VisitActivityCreate(BaseModel):
    name: str
    activity_type: str = "other"
    price: float = 0
    capacity_per_slot: int = 1
    duration_minutes: int | None = None
    requires_staff_role: str | None = None
    requires_animal_id: str | None = None
    welfare_limit_json: dict | None = None
    active: bool = True

    @field_validator("name")
    @classmethod
    def name_non_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("name cannot be empty")
        return v.strip()

    @field_validator("activity_type")
    @classmethod
    def activity_type_valid(cls, v: str) -> str:
        if v not in ACTIVITY_TYPES:
            raise ValueError(f"activity_type must be one of {sorted(ACTIVITY_TYPES)}")
        return v

    @field_validator("capacity_per_slot")
    @classmethod
    def capacity_positive(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("capacity_per_slot must be greater than zero")
        return v


class VisitActivityOut(ORMModel):
    id: str
    farm_id: str
    name: str
    activity_type: str
    price: float
    capacity_per_slot: int
    duration_minutes: int | None = None
    requires_staff_role: str | None = None
    requires_animal_id: str | None = None
    welfare_limit_json: dict | None = None
    active: bool


# ---------------------------------------------------------------------------
# Visitors
# ---------------------------------------------------------------------------
class VisitorProfileCreate(BaseModel):
    full_name: str
    phone: str | None = None
    email: str | None = None
    preferred_language: str = "en"
    notes: str | None = None
    consent_marketing: bool = False

    @field_validator("full_name")
    @classmethod
    def full_name_non_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("full_name cannot be empty")
        return v.strip()


class VisitorProfileOut(ORMModel):
    id: str
    farm_id: str
    full_name: str
    phone: str | None = None
    email: str | None = None
    preferred_language: str
    notes: str | None = None
    consent_marketing: bool


# ---------------------------------------------------------------------------
# Bookings
# ---------------------------------------------------------------------------
class BookingActivitySelection(BaseModel):
    activity_id: str
    scheduled_at: datetime
    quantity: int = 1

    @field_validator("quantity")
    @classmethod
    def quantity_positive(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("quantity must be greater than zero")
        return v


class VisitBookingCreate(BaseModel):
    visitor_id: str | None = None
    visitor: VisitorProfileCreate | None = None  # create-on-the-fly for walk-ins
    session_id: str
    package_id: str
    adults: int = 1
    children: int = 0
    activities: list[BookingActivitySelection] = []
    deposit_amount: float = 0
    source: str = "manual"
    payment_method: str | None = None
    notes: str | None = None
    idempotency_key: str | None = None

    @field_validator("adults", "children")
    @classmethod
    def guests_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("guest counts cannot be negative")
        return v

    @field_validator("source")
    @classmethod
    def source_valid(cls, v: str) -> str:
        if v not in BOOKING_SOURCES:
            raise ValueError(f"source must be one of {sorted(BOOKING_SOURCES)}")
        return v


class VisitBookingCancelRequest(BaseModel):
    reason: str | None = None
    refund: bool = False


class BookingActivityOut(ORMModel):
    id: str
    activity_id: str
    scheduled_at: datetime
    quantity: int
    unit_price: float
    status: str


class VisitBookingOut(ORMModel):
    id: str
    farm_id: str
    visitor_id: str
    session_id: str
    package_id: str
    status: str
    adults: int
    children: int
    total_amount: float
    deposit_amount: float
    balance_due: float
    source: str
    payment_method: str | None = None
    notes: str | None = None
    confirmed_at: datetime | None = None
    checked_in_at: datetime | None = None
    completed_at: datetime | None = None
    cancelled_at: datetime | None = None
    activities: list[BookingActivityOut] = []


# ---------------------------------------------------------------------------
# Staff roster + costs
# ---------------------------------------------------------------------------
class VisitStaffRosterCreate(BaseModel):
    session_id: str
    worker_id: str
    role: str
    start_time: time
    end_time: time
    hourly_rate: float = 0

    @field_validator("hourly_rate")
    @classmethod
    def hourly_rate_non_negative(cls, v: float) -> float:
        if v < 0:
            raise ValueError("hourly_rate cannot be negative")
        return v


class VisitStaffRosterOut(ORMModel):
    id: str
    session_id: str
    worker_id: str
    role: str
    start_time: time
    end_time: time
    hourly_rate: float
    total_cost: float | None = None


class VisitCostCreate(BaseModel):
    session_id: str
    category: str
    description: str | None = None
    amount: float
    allocation_method: str = "per_session"

    @field_validator("category")
    @classmethod
    def category_valid(cls, v: str) -> str:
        if v not in COST_CATEGORIES:
            raise ValueError(f"category must be one of {sorted(COST_CATEGORIES)}")
        return v

    @field_validator("allocation_method")
    @classmethod
    def allocation_method_valid(cls, v: str) -> str:
        if v not in COST_ALLOCATION_METHODS:
            raise ValueError(f"allocation_method must be one of {sorted(COST_ALLOCATION_METHODS)}")
        return v

    @field_validator("amount")
    @classmethod
    def amount_non_negative(cls, v: float) -> float:
        if v < 0:
            raise ValueError("amount cannot be negative")
        return v


class VisitCostOut(ORMModel):
    id: str
    session_id: str
    category: str
    description: str | None = None
    amount: float
    allocation_method: str


# ---------------------------------------------------------------------------
# Retail / POS
# ---------------------------------------------------------------------------
class RetailSaleLine(BaseModel):
    """Either a plain inventory item (eggs, milk, produce) or a Mouneh
    finished-goods line — never both."""

    inventory_item_id: str | None = None
    finished_goods_stock_id: str | None = None
    quantity: float
    unit_price: float

    @field_validator("quantity", "unit_price")
    @classmethod
    def positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("value must be greater than zero")
        return v


class VisitRetailSaleCreate(BaseModel):
    booking_id: str | None = None
    visitor_id: str | None = None
    channel: str = "farm_shop"
    payment_status: str = "paid"
    lines: list[RetailSaleLine]

    @field_validator("channel")
    @classmethod
    def channel_valid(cls, v: str) -> str:
        if v not in RETAIL_CHANNELS:
            raise ValueError(f"channel must be one of {sorted(RETAIL_CHANNELS)}")
        return v

    @field_validator("lines")
    @classmethod
    def lines_non_empty(cls, v: list[RetailSaleLine]) -> list[RetailSaleLine]:
        if not v:
            raise ValueError("a retail sale needs at least one line item")
        return v


class VisitRetailSaleOut(ORMModel):
    id: str
    booking_id: str | None = None
    visitor_id: str | None = None
    sale_id: str
    channel: str
    total_amount: float


# ---------------------------------------------------------------------------
# Feedback + incidents
# ---------------------------------------------------------------------------
class VisitorFeedbackCreate(BaseModel):
    booking_id: str
    rating: int
    comments: str | None = None
    would_return: bool | None = None

    @field_validator("rating")
    @classmethod
    def rating_range(cls, v: int) -> int:
        if not (1 <= v <= 5):
            raise ValueError("rating must be between 1 and 5")
        return v


class VisitorFeedbackOut(ORMModel):
    id: str
    booking_id: str
    rating: int
    comments: str | None = None
    would_return: bool | None = None
    submitted_at: datetime


class VisitIncidentCreate(BaseModel):
    session_id: str
    booking_id: str | None = None
    incident_type: str
    severity: str = "low"
    description: str
    action_taken: str | None = None

    @field_validator("incident_type")
    @classmethod
    def incident_type_valid(cls, v: str) -> str:
        if v not in INCIDENT_TYPES:
            raise ValueError(f"incident_type must be one of {sorted(INCIDENT_TYPES)}")
        return v

    @field_validator("severity")
    @classmethod
    def severity_valid(cls, v: str) -> str:
        if v not in INCIDENT_SEVERITIES:
            raise ValueError(f"severity must be one of {sorted(INCIDENT_SEVERITIES)}")
        return v


class VisitIncidentOut(ORMModel):
    id: str
    session_id: str
    booking_id: str | None = None
    incident_type: str
    severity: str
    description: str
    action_taken: str | None = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Dashboards / reports
# ---------------------------------------------------------------------------
class VisitProfitabilityOut(BaseModel):
    scope: str  # e.g. "session:<id>" or "range:2026-08-01..2026-08-31"
    package_revenue: float
    activity_revenue: float
    retail_revenue: float
    visitor_revenue: float
    staff_cost: float
    activity_cost: float
    included_product_cost: float
    cleaning_utilities_cost: float
    other_cost: float
    direct_visit_cost: float
    gross_margin: float
    checked_in_visitors: int
    revenue_per_visitor: float
    retail_conversion_pct: float
    average_basket_value: float
    by_package: list[dict]
    by_activity: list[dict]


class VisitDashboardOut(BaseModel):
    module_status: str
    upcoming_sessions: int
    today_bookings: int
    today_expected_visitors: int
    today_expected_revenue: float
    open_incidents: int
    sessions: list[VisitSessionOut]
